cpp/src/lance/arrow/dataset.cc

```diff
@@ -316,6 +317,40 @@ ::arrow::Result<std::shared_ptr<UpdaterBuilder>> LanceDataset::NewUpdate(
                                           std::move(new_field));
 }
 
+::arrow::Result<std::shared_ptr<LanceDataset>> LanceDataset::AddColumn(
+    const std::shared_ptr<::arrow::Field>& field, ::arrow::compute::Expression expression) {
+  ARROW_ASSIGN_OR_RAISE(expression, expression.Bind(*schema()));
+  auto refs = ::arrow::compute::FieldsInExpression(expression);
+  auto columns = refs | views::transform([](auto& ref) { return ref.ToString(); }) |
+                 to<std::vector<std::string>>;
+  ARROW_ASSIGN_OR_RAISE(auto builder, NewUpdate(field));
+  if (!columns.empty()) {
+    builder->Project(columns);
+  }
+  ARROW_ASSIGN_OR_RAISE(auto updater, builder->Finish());
+
+  while (true) {
+    ARROW_ASSIGN_OR_RAISE(auto batch, updater->Next());
+    if (!batch) {
+      break;
+    }
+
+    ARROW_ASSIGN_OR_RAISE(auto datum,
+                          ::arrow::compute::ExecuteScalarExpression(expression, *schema(), batch));
+    std::shared_ptr<::arrow::Array> arr;
+    if (datum.is_scalar()) {
```

Comment:
ok so this is a constant literal value?

Reply:
Yes, this is for case like `AddColumn(field, pc::literal(1234))`.